proposals/LP0004-project-governance.md

```diff
@@ -0,0 +1,393 @@
+
+## Impact On Other Projects
+
+This proposal impacts all LLVM projects and all contributors to the LLVM community.
+
+## Frequently Asked Questions
+
+If there were common questions that came up in the pitch phase, please summarize them and what you think the answer is.  Doing so gives you a chance to address them before the formal review starts up, which can lead to a more productive review thread - one that spends less time rehashing points that have already come up.
+
+## Appendix 1: LLVM Community Metrics Methods
+
+This document references some LLVM community metrics. This appendix contains the methodology and tools used to produce the metrics cited above. The metrics are gathered from git metadata with a scoped date range of September 1, 2022 through September 1, 2023. Using a specific date range allows these metrics to be re-generated consistently, however the metrics should be extended to encompass the latest usage patterns when driving decision making.
+
+### git-crawler.py
+
+This script traverses the directories in the LLVM project listing all the authors of commits in the directory once per commit. The script sums up the number of times an authors name appears to produce an approximation of the number of unique authors who have contributed code under a specific sub-directory. The script does not traverse into sub-directories that have less than 50 unique authors, and it skips `test` and `unittests` folders.
+
+The output printed is the number of unique authors that contribute to each directory with 50 or more authors, and a list of the top ten authors that committed to that directory with their number of commits.
+
+```python
+#!/usr/bin/env python3
+
+import os
+import subprocess
+from collections import Counter
+
+def handleDirectory(path):
+ path = path.replace(os.getcwd() + '/', ':')
+ output = subprocess.check_output(['git', 'log', '--format=%an', '--since=September 1 2022', '--until=September 1 2023', '--', path]).decode('utf-8').splitlines()
+ authors = {}
+ for line in output:
+   if line not in authors:
+     authors[line] = 0
+   authors[line] += 1
+ return authors
+
+results = {}
+results[os.getcwd()] = handleDirectory(os.getcwd())
+i = 0
+for root, dirs, files in os.walk(os.getcwd()):
+ i += 1
+ if root not in results:
+   dirs = []
+ toRemove = []
+ for dir in dirs:
+   if dir == 'test' or dir == 'unittests':
+     toRemove.append(dir)
+     continue
+
+   path = os.path.join(root, dir)
+   results[path] = handleDirectory(path)
+   if len(results[path]) < 50:
+     toRemove.append(dir)
+   else:
+     print('%s: %d' % (path, len(results[path])) )
+ for dir in toRemove:
+   dirs.remove(dir)
+
+for dir in results:
+ if not results[dir]:
+   continue
+ counters = Counter(results[dir])
+ high = counters.most_common(10)
+ print('%s:' % dir)
+ for val in high:
+   print('  %s: %d' % (str(val[0]), val[1])) \
+````
```

Comment:
truncated content? Also 3 backquotes

Reply:
When removed the trailing `\` I got:
```
~/llvm-project/lldb$ python3 /tmp/foo.py
/home/david.spickett/llvm-project/lldb/include: 53
/home/david.spickett/llvm-project/lldb/source: 121
/home/david.spickett/llvm-project/lldb/include/lldb: 53
/home/david.spickett/llvm-project/lldb/source/Plugins: 103
/home/david.spickett/llvm-project/lldb:
  Alex Langford: 174
  Jonas Devlieghere: 159
  Med Ismail Bennani: 141
```
So seems to work fine but maybe missing some intended output.

Reply:
I'll fix the formatting. We were iterating on the document in Google docs and some things got messed up in the conversion to markdown. The full script is there the trailing `\` seems to have just been inserted by the markdown converter.